pkg/cmd/experimental/diagnostics/pod.go

```diff
@@ -0,0 +1,193 @@
+	failed := false
+	warnings := []error{}
+	errors := []error{}
+	diagnostics := []types.Diagnostic{}
+
+	func() { // don't trust discovery/build of diagnostics; wrap panic nicely in case of developer error
+		defer func() {
+			if r := recover(); r != nil {
+				failed = true
+				stack := debug.Stack()
+				errors = append(errors, fmt.Errorf("While building the diagnostics, a panic was encountered.\nThis is a bug in diagnostics. Error and stack trace follow: \n%v\n%s", r, stack))
+			}
+		}() // deferred panic handler
+		podDiags, ok, err := o.buildPodDiagnostics()
+		failed = failed || !ok
+		if ok {
+			diagnostics = append(diagnostics, podDiags...)
+		}
+		if err != nil {
+			errors = append(errors, err...)
+		}
+
+	}()
+
+	if failed {
+		return failed, kutilerrors.NewAggregate(errors), len(warnings), len(errors)
+	}
+
+	failed, err, numWarnings, numErrors := runDiagnostics(o.Logger, diagnostics, len(warnings), len(errors))
+	return failed, err, numWarnings, numErrors
+}
+
+// runDiagnostics performs the actual execution of diagnostics once they're built.
+func runDiagnostics(logger *log.Logger, diagnostics []types.Diagnostic, warnCount int, errorCount int) (bool, error, int, int) {
+	for _, diagnostic := range diagnostics {
+		func() { // wrap diagnostic panic nicely in case of developer error
+			defer func() {
+				if r := recover(); r != nil {
+					errorCount += 1
+					stack := debug.Stack()
+					logger.Error("CED5001",
+						fmt.Sprintf("While running the %s diagnostic, a panic was encountered.\nThis is a bug in diagnostics. Error and stack trace follow: \n%s\n%s",
+							diagnostic.Name(), fmt.Sprintf("%v", r), stack))
+				}
+			}()
+
+			if canRun, reason := diagnostic.CanRun(); !canRun {
+				if reason == nil {
+					logger.Notice("CED5002", fmt.Sprintf("Skipping diagnostic: %s\nDescription: %s", diagnostic.Name(), diagnostic.Description()))
+				} else {
+					logger.Notice("CED5003", fmt.Sprintf("Skipping diagnostic: %s\nDescription: %s\nBecause: %s", diagnostic.Name(), diagnostic.Description(), reason.Error()))
+				}
+				return
+			}
+
+			logger.Notice("CED5004", fmt.Sprintf("Running diagnostic: %s\nDescription: %s", diagnostic.Name(), diagnostic.Description()))
+			r := diagnostic.Check()
+			for _, entry := range r.Logs() {
+				logger.LogEntry(entry)
+			}
+			warnCount += len(r.Warnings())
+			errorCount += len(r.Errors())
+		}()
+	}
+	return errorCount > 0, nil, warnCount, errorCount
+}
+
+var (
+	// availablePodDiagnostics contains the names of host diagnostics that can be executed
+	// during a single run of diagnostics. Add more diagnostics to the list as they are defined.
+	availablePodDiagnostics = sets.NewString(poddiag.PodCheckDnsName, poddiag.PodCheckAuthName)
+)
+
+// buildPodDiagnostics builds host Diagnostic objects based on the host environment.
+// Returns the Diagnostics built, "ok" bool for whether to proceed or abort, and an error if any was encountered during the building of diagnostics.) {
```

Comment:
Fix the end of the comment. Also it would be nice to have this function before `runDiagnostics` since it appears earlier in its caller.